File: python3/is_graph_bipartite.py
import collections

"""
Space : O(n)
Time  : O(n^2)

BFS approach
Method:
Graph coloring

alternatively you can use deque
for better pop time performance
"""


class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        """
        :type graph: List[List[int]]
        :rtype: bool
        """
        n, colored = len(graph), {}
        for i in range(n):
            if i not in colored and graph[i]:
                colored[i] = 1
                q = collections.deque([i])
                while q:
                    cur = q.popleft()
                    for nb in graph[cur]:
                        if nb not in colored:
                            colored[nb] = -colored[cur]
                            q.append(nb)
                        elif colored[nb] == colored[cur]:
                            return False
        return True

File: golang/is_graph_bipartite.go
package main

/*
Space	: O(n)
Time	: O(n^2)
*/

func isBipartite(graph [][]int) bool {
	n := len(graph)
	colored := map[int]int{}

	for i := 0; i < n; i++ {
		_, ok := colored[i]
		if !ok && len(graph[i]) > 0 {
			colored[i] = 1
			q := []int{i}
			for len(q) > 0 {
				cur := q[0]
				q = q[1:]
				for _, nb := range graph[cur] {
					if _, ok := colored[nb]; !ok {
						colored[nb] = -colored[cur]
						q = append(q, nb)
					} else if colored[nb] == colored[cur] {
						return false
					}
				}
			}
		}
	}

	return true
}
